### ISSUE TITLE
Segmentation fault when running SPEC 2017 benchmark

### ISSUE THREAD

**Issue by CodeATA** [technion-csl/mosmodel#16]
I'm running SPEC 2017 benchmarks to build their performance models. Most of the benchmarks finished successfully, but I encountered a segmentation fault for 605.mcf_s.

```
========== [INFO] start producing: experiments/single_page_size/layout2mb/repeat1/perf.out ==========
creating a new output directory /home/chendongwei/tlb-workspace/perf-models/605.mcf/experiments/single_page_size/layout2mb/repeat1/ ...
copying the benchmark files to /home/chendongwei/tlb-workspace/perf-models/605.mcf/experiments/single_page_size/layout2mb/repeat1/...
warming up before running...
running the benchmark /home/chendongwei/tlb-workspace/perf-models/605.mcf/benchmark...
the full submit command is:
	/home/chendongwei/tlb-workspace/perf-models/605.mcf/scripts/measureGeneralMetrics.sh /home/chendongwei/tlb-workspace/perf-models/605.mcf/scripts/setCpuMemoryAffinity.sh 1 	/home/chendongwei/tlb-workspace/perf-models/605.mcf/mosalloc/runMosalloc.py --library /home/chendongwei/tlb-workspace/perf-models/605.mcf/mosalloc/src/libmosalloc.so  -fps 1GB -aps 104857600 -as1 0 -ae1 0 -as2 0 -ae2 104857600 -bps 4259315712 -bs1 0 -be1 0 -bs2 0 -be2 4259315712 --analyze ./run.sh
waiting for the run to complete...
Traceback (most recent call last):
  File "/home/chendongwei/tlb-workspace/perf-models/605.mcf/scripts/runBenchmark.py", line 106, in <module>
    benchmark_run.wait()
  File "/home/chendongwei/tlb-workspace/perf-models/605.mcf/scripts/runBenchmark.py", line 54, in wait
    raise subprocess.CalledProcessError(self._run_process.returncode, ' '.join(self._run_process.args))
subprocess.CalledProcessError: Command '/home/chendongwei/tlb-workspace/perf-models/605.mcf/scripts/measureGeneralMetrics.sh /home/chendongwei/tlb-workspace/perf-models/605.mcf/scripts/setCpuMemoryAffinity.sh 1 /home/chendongwei/tlb-workspace/perf-models/605.mcf/mosalloc/runMosalloc.py --library /home/chendongwei/tlb-workspace/perf-models/605.mcf/mosalloc/src/libmosalloc.so -fps 1GB -aps 104857600 -as1 0 -ae1 0 -as2 0 -ae2 104857600 -bps 4259315712 -bs1 0 -be1 0 -bs2 0 -be2 4259315712 --analyze ./run.sh' returned non-zero exit status 139.
/home/chendongwei/tlb-workspace/perf-models/605.mcf/experiments/template.mk:36: recipe for target 'experiments/single_page_size/layout2mb/repeat1/perf.out' failed
make: *** [experiments/single_page_size/layout2mb/repeat1/perf.out] Error 1
```

I uploaded some relevant files, including the gdb backtrace log after the segfault happens.

[mcf debug.zip](https://github.com/technion-csl/mosmodel/files/7083886/mcf.debug.zip)

**Comment by CodeATA** [comment]
Get a similar error running graph500 (using https://github.com/graph500/graph500). Relevant files are here.

[graph500 debug.zip](https://github.com/technion-csl/mosmodel/files/7090353/graph500.debug.zip)

Thanks.

**Comment by idanyani** [comment]
Regarding graph500: it may be a bug in the benchmark.
You log seems similar to this bug report (which hasn't been fixed yet, unfortunately):
https://github.com/graph500/graph500/issues/10

Regarding spec17/605.mcf_s: we will try to reproduce the error on our side but it may not work because our system is a bit different. For example, our SPEC CPU version is 1.1.8 and not 1.11.

Meanwhile, can you please run these two benchmark under "valgrind --leak-check=full" **without mosalloc** to make sure that they don't contain any memory errors in their baseline implementation?

**Comment by CodeATA** [comment]
Valgrind reported no leak on mcf.
As for graph500, somehow Valgrind could not make it run and gave the following messages.
I checked the bug report of graph500. I can successfully run the benchmark without mosalloc, so I don't think they are the same bug.

```
==3257== Memcheck, a memory error detector
==3257== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==3257== Using Valgrind-3.11.0 and LibVEX; rerun with -h for copyright info
==3257== Command: ./src/graph500_reference_bfs 25 16
==3257== 
==3257== Warning: ignored attempt to set SIGRT32 handler in sigaction();
==3257==          the SIGRT32 signal is used internally by Valgrind
cr_libinit.c:189 cri_init: sigaction() failed: Invalid argument
==3257== 
==3257== Process terminating with default action of signal 6 (SIGABRT)
==3257==    at 0x5820438: raise (raise.c:54)
==3257==    by 0x5822039: abort (abort.c:89)
==3257==    by 0x5BB813A: ??? (in /usr/lib/libcr.so.0.5.5)
==3257==    by 0x40106F9: call_init.part.0 (dl-init.c:72)
==3257==    by 0x401080A: call_init (dl-init.c:30)
==3257==    by 0x401080A: _dl_init (dl-init.c:120)
==3257==    by 0x4000C69: ??? (in /lib/x86_64-linux-gnu/ld-2.23.so)
==3257==    by 0x2: ???
==3257==    by 0xFFF000472: ???
==3257==    by 0xFFF00048F: ???
==3257==    by 0xFFF000492: ???
==3257== 
==3257== HEAP SUMMARY:
==3257==     in use at exit: 0 bytes in 0 blocks
==3257==   total heap usage: 1 allocs, 1 frees, 25 bytes allocated
==3257== 
==3257== All heap blocks were freed -- no leaks are possible
==3257== 
==3257== For counts of detected and suppressed errors, rerun with: -v
==3257== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
Aborted (core dumped)
```

**Comment by idanyani** [comment]
Hi @CodeATA ,

I managed to reproduce your bug in spec_cpu2017/605.mcf_s. We will check it asap.

Regarding graph500: we need a few details please.
What version of the benchmark are you using? You can refer us to the tarball that you downloaded or to the commit hash of the official github repo https://github.com/graph500/graph500
What compiler did you use to build it?

**Comment by CodeATA** [comment]
I'm using the graph500-3.0.0 tarball, commit hash 9dbb76c.
The compiler is mpicc from MPICH 3.2 with gcc 5.4.0

```
~$ mpicc -v
mpicc for MPICH version 3.2
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/5/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.4.0-6ubuntu1~16.04.12' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.12) 
```

**Comment by magbarya** [comment]
Debugging the failures shows that the brk pool size is not large enough to serve all allocation requests.
The brk pool size is determined by running the benchmark and measuring its RSS using time utility.

The time utility reports the max RSS, in other words, the maximal set size of the physical memory consumed (allocated and accessed) by the benchmark. The Mosalloc brk pool increases each time the glibc calls the morecore to serve an allocation request regardless if that allocation will be really used (accesses) by the benchmark or not. Since the heap is lazily allocated, only accessed pages will be counted by the max-RSS. For example, the mcf benchmark allocates ~5GB virtual memory but accesses only ~4GB of these allocations. The time utility will report for ~4GB as the max-RSS and mosmodel will use that amount of memory to set the brk pool size, which will be less than what mcf allocates.

In other words, Mosalloc requires to set the brk pool size to the anonymous max-VSZ (virtual set size).  Thus, instead of running the benchmark under time to measure its memory footprint, which could be less than what is required---RSS<VSZ, the mosmodel code uses Mosalloc (by configuring it to set the brk pool size to RAM size) to measure the exact size of the brk pool.

The following commits include the fix: [b341175](b341175b17451d15de433b8e5e1d6929d012dc4f) and [018592f](018592f7bf536240d21298cfcd17122c1382bd94)
The fix was tested on spec2017/mcf and graph500v.3/reference_bfs

**Comment by magbarya** [comment]
Hi @CodeATA,
First of all, I would like to thank you for catching this bug. I can confirm that this bug is in Mosmodel code and it was already fixed and tested.
I will appreciate it if you could pull the latest changes from the master branch (or commit: [018592f](018592f7bf536240d21298cfcd17122c1382bd94)), run "make clean" and "make" again (or to clone a clean repository and run it again), and check if that fixes the issue from your side.

Thanks.

**Comment by CodeATA** [comment]
Sorry for the late reply. I have tried the two benchmarks. mcf is doing fine, but graph500 still doesn't work.

graph500 reports "failed to allocate memory by mmap" when running growing_window_2m. Here is the benchmark.log

[benchmark.log](https://github.com/technion-csl/mosmodel/files/7256813/benchmark.log)

**Comment by magbarya** [comment]
Can you please add more details about how are you running the graph500 benchmark? which kernel are you using (bfs/sssp) and what scale? more specifically, can you please provide us with the full command line you are running so we can try to reproduce this issue in our machine?

**Comment by CodeATA** [comment]
I'm using the bfs kernel at scale 25 (./graph500_reference_bfs 25 16). I tried using scale 10 and got the same result.

The command line in run.sh is ```./src/graph500_reference_bfs 10 16```. I use ```make analysis/mosmodel``` to build the model.

Please let me know if you need more details. Thanks.

**Comment by magbarya** [comment]
@CodeATA, can you please provide the failure log? at which stage/experiment does it fail?
I have tried to run graph500_reference_bfs at scale=25 without mosalloc/mosmodel and it fails with segmentation fault (i.e., the benchmark is broken). Running it at scale=10 works fine for me. Can you please try to run it at scale=25 without mosalloc/mosmoel and see if it works smoothly?

**Comment by CodeATA** [comment]
I encountered the same segmentation fault after a recompilation on another machine. For now, I could not figure out why it worked before. And I checked the log at scale=10, I think it is the same issue as the old gups toy benchmark. I'm rerunning graph500 with scale=23, which passes the segmentation fault point without mosalloc. I'll let you know if any result comes out.

**Comment by CodeATA** [comment]
mosmodel finished properly at scale=23. I'm closing this issue.
